### PR TITLE
[Performance] Fix array_merge in a loop

### LINKED ISSUE #33929: [Issue] [Performance] Fix array_merge in a loop
This issue is automatically created based on existing pull request: magento/magento2#33904: [Performance] Fix array_merge in a loop

---------

### Fixed Issues (if relevant)

1. Fixes magento/magento2#<issue_number>

### Manual testing scenarios (*)

1. ...
2. ...

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### PR DESCRIPTION
**Pull request by ihor-sviziev**
### Fixed Issues (if relevant)

1. Fixes magento/magento2#<issue_number>

### Manual testing scenarios (*)

1. ...
2. ...

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#33929: [Performance] Fix array_merge in a loop

### CODE REVIEW

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Review by andrewbess — CHANGES_REQUESTED** [review]
Hello @ihor-sviziev 
Thank you for your contribution.
Great job.
Could you please update branch?
Thank you in advance.

**Review by dudzio12 — APPROVED** [review]

File: setup/src/Magento/Setup/Module/Di/App/Task/Operation/ApplicationCodeGenerator.php [hunk 1]
```diff
@@ -5,7 +5,6 @@
  */
 namespace Magento\Setup\Module\Di\App\Task\Operation;
 
-use Magento\Framework\App\Bootstrap;
```
Comment (tanghongtanghong) [inline setup/src/Magento/Setup/Module/Di/App/Task/Operation/ApplicationCodeGenerator.php:8]: test

**Review by Den4ik — APPROVED** [review]
@ihor-sviziev Good job 👍

**Comment by engcom-Alfa** [comment]
Hi @ihor-sviziev Thanks for your collaboration and contribution!
As part of QA process, I will be just require the steps to execute the to validate the performance improvements.
Could you please help me providing detailed steps to reproduce the case?

**Comment by ihor-sviziev** [comment]
Hi @engcom-Alfa,
This is a performance improvement for the code generation step. It should positively affect the `setup:di:compile` time on magneto installations with a lot of extensions

**Comment by engcom-Alfa** [comment]
Hi @ihor-sviziev , @tanghongtanghong , @andrewbess , @Den4ik , @dudzio12 
Thanks for the collaboration & contribution!

:x: QA not Passed

**_di:compile_** command performance needs to be reduced 

**Manual Scenario Steps**

1. Installed a fresh Magento 2.4-develop and installed a couple of extensions eg- Braintree and SMTP
2. Upgraded the Magento and make sure that the extensions are installed properly.
3. Executed the di:compile command and noted the execution time.


**:heavy_check_mark: Expected result After Fix** 
- Currently the execution time of the command taking 40 seconds with the extensions; Expected to reduce the execution time ~10% at least.


**:x: Actual result After Fix** 
- After applying the PR changes; the execution time is being same as ~40 seconds;
  Screenshot 2021-11-09 at 4 55 05 PM

Kindly take a look at above details and recommend if there is any information will help by any chance please!

**Review by engcom-Alfa — CHANGES_REQUESTED** [review]
Requesting changes if required as per above comment
Thanks

**Comment by Den4ik** [comment]
@engcom-Alfa It's don't provide huge performance increase, but this PR decrease count of merge function calls that is good point for me.

2021-11-10-16-18-49@2x

**Comment by engcom-Dash** [comment]
> @engcom-Alfa It's don't provide huge performance increase, but this PR decrease count of merge function calls that is good point for me.
> 
> 2021-11-10-16-18-49@2x

Hi @Den4ik  
In order to count the total number of functions call which tool you are using. Could you please suggest so that we can use the same.

**Comment by ihor-sviziev** [comment]
@engcom-Dash, this is https://www.blackfire.io/

**Comment by Den4ik** [comment]
Hi @engcom-Dash 
It's blackfire https://blackfire.io/

**Comment by engcom-Dash** [comment]
:heavy_check_mark: QA Passed

**Manual testing scenario:**

1. Run the blackfire profiler on magento2.4-develop with command ‘blackfire run bin/magento se:di:co’
2. Run the blackfire profiler on PR barnch with command ‘blackfire run bin/magento se:di:co’
3. compare both the profile on blackfire.
 
**Before:** :heavy_multiplication_x:   The count of array merge function is high

Screenshot 2022-01-11 at 5 28 17 PM

**After:** :heavy_check_mark:  The count of array merge function has been significantly reduced.

Screenshot 2022-01-11 at 5 27 29 PM (1)

FYI .. To Configure blackfire in system follow the below link:
https://blackfire.io/docs/up-and-running/installation

**Comment by engcom-Charlie** [comment]
It seems that bot moved this PR to Changes Requested from  Merge in Progress, so moving it back.

Thank you!

**Pull request merged**